ext/odbc/php_odbc.c

```diff
@@ -49,23 +51,175 @@
+	} else {
+		close_results_with_connection(link->connection);
+
+		zend_hash_del(&ODBCG(connections), link->hash);
+
+		/* If aborted via timer expiration, don't try to call any unixODBC function */
+		if (!(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+			safe_odbc_disconnect(link->connection->hdbc);
+			SQLFreeConnect(link->connection->hdbc);
+			SQLFreeEnv(link->connection->henv);
+		}
+		efree(link->connection);
+		ODBCG(num_links)--;
+	}
+
+	link->connection = NULL;
+
+	zend_string_release(link->hash);
+	link->hash = NULL;
+}
+
+static void odbc_connection_free_obj(zend_object *obj)
+{
+	odbc_link *link = odbc_link_from_obj(obj);
+
+	if (!link->connection) {
+		zend_object_std_dtor(&link->std);
+		return;
+	}
+
+	odbc_link_free(link);
+	zend_object_std_dtor(&link->std);
+}
+
+static inline odbc_result *odbc_result_from_obj(zend_object *obj) {
+	return (odbc_result *)((char *)(obj) - XtOffsetOf(odbc_result, std));
+}
+
+#define Z_ODBC_RESULT_P(zv) odbc_result_from_obj(Z_OBJ_P(zv))
+
+static zend_object *odbc_result_create_object(zend_class_entry *class_type) {
+	odbc_result *intern = zend_object_alloc(sizeof(odbc_result), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &odbc_result_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *odbc_result_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct ODBC\\Result, use an appropriate odbc_* function instead");
+	return NULL;
+}
+
+static void odbc_result_free(odbc_result *res)
+{
+	ZEND_ASSERT(res->conn_ptr && "result has already been closed");
+
+	if (res->values) {
+		for (int i = 0; i < res->numcols; i++) {
+			if (res->values[i].value) {
+				efree(res->values[i].value);
+			}
+		}
+		efree(res->values);
+		res->values = NULL;
+		res->numcols = 0;
+	}
+	/* If aborted via timer expiration, don't try to call any unixODBC function */
+	if (res->stmt && !(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+#if defined(HAVE_SOLID) || defined(HAVE_SOLID_30) || defined(HAVE_SOLID_35)
+		SQLTransact(res->conn_ptr->henv, res->conn_ptr->hdbc,
+					(SQLUSMALLINT) SQL_COMMIT);
+#endif
+		SQLFreeStmt(res->stmt,SQL_DROP);
+		/* We don't want the connection to be closed after the last statement has been closed
+		 * Connections will be closed on shutdown
+		 */
+		res->stmt = NULL;
+	}
+	if (res->param_info) {
+		efree(res->param_info);
+		res->param_info = NULL;
+	}
+	res->conn_ptr = NULL;
+}
 
-static int le_result, le_conn, le_pconn;
+static void odbc_result_free_obj(zend_object *obj) {
+	zend_ulong num_index;
+	zval *p;
+
+	odbc_result *result = odbc_result_from_obj(obj);
+
+	if (!result->conn_ptr) {
+		zend_object_std_dtor(&result->std);
+		return;
+	}
+
+	ZEND_HASH_FOREACH_NUM_KEY_VAL(&ODBCG(results), num_index, p) {
+		odbc_result *res = Z_ODBC_RESULT_P(p);
+		if (res->stmt == result->stmt) {
+			zend_hash_index_del(&ODBCG(results), num_index);
```

Comment:
Yeah, I was wondering about the same when implementing these foreach based modifications. I'll try to look your changes up and fix accordingly. Since a statement should only be present once in the hash table, I guess the fix is going to be easy.

Reply:
If it only occurs once, then you can the removal it in the loop, but use a `break;` when the item is removed.
This is what I was referring to: https://github.com/php/php-src/commit/7e4a3236d92bc4bcad9ad6661d96f3da92d9c4d3
Although it might be fine in your case as the hashtable is not resized.